### ISSUE TITLE
Perl compilation failing after removing berkeley-db dependency [cannot find -ldb] on CentOS7

### ISSUE THREAD

**Issue by thiagogenez** [Homebrew/linuxbrew-core#22314]
https://github.com/Homebrew/linuxbrew-core/blob/80c2e0f0ff24f9327233a785645c3345372eb4d9/Formula/perl.rb#L42-L43

Dependencies below removed during the merge of commit `64cee040948c37ad526060099b13af65f565c6a5 `
https://github.com/Homebrew/linuxbrew-core/blob/51e2cd9daa514846c6ba781d23591a48b9c505f2/Formula/perl.rb#L24-L27

```
==> ./Configure -des -Dprefix=/home/thiagogenez_ebi_ac_uk/.linuxbrew/Cellar/perl/5.32.1 -Dprivlib=/home/thiagogenez_ebi_ac_uk/.linuxbrew/Cellar/perl/5.32.1/lib/perl5/5.32.1 -Dsitelib=/home/thiagogenez_ebi_ac_uk/.linuxbrew/Cellar/perl/5.32.1/lib/perl5/site_perl/5.32.1 -Dotherlibdirs=/home/thiagogenez_ebi_ac_uk/.linuxbrew/lib/perl5/site_perl/5.32.1 -Dperlpath=/home/thiagogenez_ebi_ac_uk/.linuxbrew/opt/perl/bin/perl -Dstartperl=#!/home/thiagogenez_ebi_ac_uk/.linuxbrew/opt/perl/bin/perl -Dman1dir=/home/thiagogenez_ebi_ac_uk/.linuxbrew/Cellar/perl/5.32.1/share/man/man1 -Dman3dir=/home/thiagogenez_ebi_ac_uk/.linuxbrew/Cellar/perl/5.32.1/share/man/man3 -Duseshrplib -Duselargefiles -Dusethreads -Dnoextensions=GDBM_File -Dnoextensions=DB_File -Dlocincpth=/home/thiagogenez_ebi_ac_uk/.linuxbrew/include
First let's make sure your kit is complete.  Checking...
Locating common programs...
Checking compatibility between /usr/bin/echo and builtin echo (if any)...
Symbolic links are supported.
Checking how to test for symbolic links...
You can test for symbolic links with 'test -h'.
Checking for cross-compile
No targethost for running compiler tests against defined, running locally
Good, your tr supports [:lower:] and [:upper:] to convert case.
Using [:upper:] and [:lower:] to convert case.
aix_3         dragonfly       irix_5   		nonstopux     sunos_4_0
aix_4         dynixptx        irix_6_0   	openbsd       sunos_4_1
aix   	      dynix   	      irix_6_1   	opus          super-ux
altos486      epix   	      irix_6   		os2   	      svr4
amigaos       esix4   	      isc_2   		os390         svr5
atheos        fps   	      isc   		os400         ti1500
aux_3         freebsd         linux-android   	posix-bc      ultrix_4
bitrig        freemint        linux   		qnx   	      umips
bsdos         gnukfreebsd     lynxos   		riscos        unicosmk
catamount     gnuknetbsd      midnightbsd   	sco_2_3_0     unicos
convexos      gnu   	      minix   		sco_2_3_1     unisysdynix
cxux          greenhills      mips   		sco_2_3_2     utekv
cygwin        haiku   	      mirbsd   		sco_2_3_3     uwin
darwin        hpux   	      mpc   		sco_2_3_4     vos
dcosx         i386   	      ncr_tower   	sco
dec_osf       interix         netbsd   		solaris_2
dos_djgpp     irix_4          newsos4   	stellar
Which of these apply, if any? [linux]
Operating system name? [linux]
Operating system version? [3.10.0-1160.11.1.el7.x86_64]
Installation prefix to use? (~name ok)
[/home/thiagogenez_ebi_ac_uk/.linuxbrew/Cellar/perl/5.32.1]
AFS does not seem to be running...
What installation prefix should I use for installing files? (~name ok)
[/home/thiagogenez_ebi_ac_uk/.linuxbrew/Cellar/perl/5.32.1]
Getting the current patchlevel...
Build a threading Perl? [y]
Use which C compiler? [cc]
Checking for GNU cc in disguise and/or its version number...
Now, how can we feed standard input to your C preprocessor...
Directories to use for library searches?
[/home/thiagogenez_ebi_ac_uk/.linuxbrew/lib /home/thiagogenez_ebi_ac_uk/.linuxbrew/Cellar/gcc@5/5.5.0_6/bin/../lib/gcc/5/gcc/x86_64-unknown-linux-gnu/5.5.0/include-fixed /home/thiagogenez_ebi_ac_uk/.linuxbrew/Cellar/gcc@5/5.5.0_6/bin/../lib/gcc/5/gcc/../../../../lib /usr/local/lib /lib/../lib64 /usr/lib/../lib64 /lib /usr/lib /lib64 /usr/lib64 /usr/local/lib64]
What is the file extension used for shared libraries? [so]
Make shared library basenames unique? [n]
Build Perl for SOCKS? [n]
Try to use long doubles if available? [n]
Checking for optional libraries...
What libraries to use?
[-lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat]
What optimizer/debugger flag should be used? [-O2]
Any additional cc flags?
[-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/home/thiagogenez_ebi_ac_uk/.linuxbrew/include]
Let me guess what the preprocessor flags are...
Any additional ld flags (NOT including libraries)?
[ -fstack-protector-strong -L/usr/local/lib]
Checking your choice of C compiler and flags for coherency...
I've tried to compile and run the following simple program:

#include <stdio.h>
int main() { printf("Ok\n"); return(0); }

I used the command:

	cc -o try -O2 -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/home/thiagogenez_ebi_ac_uk/.linuxbrew/include -fstack-protector-strong -L/usr/local/lib try.c -lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
	 ./try

and I got the following output:

/home/thiagogenez_ebi_ac_uk/.linuxbrew/bin/ld: cannot find -ldb
collect2: error: ld returned 1 exit status
I can't compile the test program.
You have a BIG problem.  Shall I abort Configure [y]
Ok.  Stopping Configure.
/usr/bin/curl --version
/usr/bin/ruby -e puts\ RUBY_VERSION
/home/thiagogenez_ebi_ac_uk/.linuxbrew/Homebrew/Library/Homebrew/debrew.rb:18:in `raise'
FormulaUnavailableError: No available formula with the name "xorg".
```

**Comment by danielnachun** [comment]
I've reproduced in the Homebrew CI Docker environment, so it's not specific to your setup.  I'm looking into this now.

**Comment by thiagogenez** [comment]
Thanks @danielnachun 

I was investingating this tonight using a Docker environment. I realised this problem is on CentOS 7 with Perl's configure script.

On CentOS container:
```
Checking for optional libraries...
What libraries to use?[-lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat]
```

On Ubuntu container:
```
Checking for optional libraries...
What libraries to use? [-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc]
```
It seems `Configure` is ignoring the flags `-Dnoextensions=GDBM_File` and `-Dnoextensions=DB_File` when on CentOs7.

**Comment by danielnachun** [comment]
We merged a commit I made earlier this week to fix a completely unrelated problem: https://github.com/Homebrew/brew/pull/10606 and that seems to be the cause of this failure.  I'm hoping to fix this without having to completely revert that commit, but if not I'll just get it reverted.

**Comment by danielnachun** [comment]
I got the problematic PR reverted and Perl built from source in our CI Docker environment.  Please try again as I believe it should be working now.

**Comment by danielnachun** [comment]
This appears to be fixed now.

**Comment by thiagogenez** [comment]
Thanks @danielnachun !

**Comment by atwok** [comment]
Noob question.  Apologies.

How do I make sure I have the fix for this?  I just tried to install curl on a Centos 7 box and hit this error while trying to install Perl as a dependency.

Thanks!

**Comment by thiagogenez** [comment]
Weirdly I hit again on the same error message after started trying a brand-new brew environment.

@atwok, try `brew install berkeley-db` before `brew install perl`

**Comment by danielnachun** [comment]
I'm reopening this to test again.

**Comment by danielnachun** [comment]
I can't reproduce this in either the Docker CI or CentOS 7.  Can you both share the output of `brew doctor` and `brew config`?  It's possible you are somehow still using an older `brew` which still has the bad commit.

**Comment by ankitkk** [comment]
I ran into this with a fresh install  moments ago - because brew is not installed yet, there's no way to  install berkley-db before perl.

**Comment by danielnachun** [comment]
Ugh, I was able to reproduce this in CentOS 7 finally and started digging into why it was still trying to link to `berkeley-db` and `gdbm` and it's not pretty.  

It appears that, in addition to using those libraries for the XS extensions (Perl extensions which include native compiled code), the Perl build system also wants to link to them while building the core Perl binaries.  

@iMichka I found in the `Configure` script for perl that it hardcodes (unwisely, IMO) the variable `libswanted` with the following values:

```
libswanted="cl pthread socket bind inet nsl ndbm gdbm dbm db malloc dl ld"
libswanted="$libswanted sun m crypt sec util c cposix posix ucb bsd BSD"
```

I can override this variable by adding `-A define:libswanted`.  I removed `ndbm`, `gdbm`, `dbm`, `db` and that stopped it from trying to link to `gdbm` and `db`.  I also removed `sun` and `ucb` because it's 2021 and it's time to move on.  I also found that I needed to remove `bsd` and `BSD`, which is an odd thing to have to do on Linux, but `libbsd` is a thing and it tried to link to my brewed copy of that, which means it could do the same for a system installation as well.
 
This was the actual argument I add to the Configure args:
```
      -A define:libswanted='cl pthread socket bind inet nsl malloc dl ld m crypt sec util c cposix posix'
```

Lastly, I wanted to check if `DB_File` and `GDBM_File` could still be installed even if the core Perl binaries weren't linked to the shared libraries for those respective databases.  `DB_File` installed just fine.  The real kicker though is that `GDBM_File` is not available as a standalone CPAN module, only as part of the main Perl source.  It can only be built by reinstalling Perl itself!  I can imagine that causing some grief for users.

It's worth noting at this point that even the famously minimalist Arch Linux builds Perl with GDBM and BerkeleyDB support: https://github.com/archlinux/svntogit-packages/blob/packages/perl/trunk/PKGBUILD.  I was looking into why our BerkeleyDB installation is so large compared to the one for Arch Linux (140MB in Homebrew vs 6MB in Arch Linux), and I see that 120MB of that is documentation 👎 and it also contains static copies of all the libraries that we will never use.

I am now thinking the best option here may be to re-enable GDBM and BerkeleyDB support and clean up the BerkeleyDB formula so that it doesn't include so much unneeded cruft.  We may also have to add `libbsd` as a dependency, since there is no easy way to disable that from trying to link to it aside from overriding `libswanted`, and it is a small relocatable bottle.

@ankitkk @thiagogenez Thanks for following up on this.  As you can see this has turned to be a much more complex issue than I originally though it would be.  Once my fellow maintainers and I decide how best to handle this, the fix for this should be really simple.

**Comment by danielnachun** [comment]
One quick update: I was able to make trivial changes to `berkeley-db` to reduce size in the cellar to ~7 MB, a massive improvement over what we had before.

**Comment by iMichka** [comment]
Looks like we have no choice then. Let's add both dependencies back. Can you open a PR in homebrew-core directly?

**Comment by thiagogenez** [comment]
Thanks, @danielnachun for digging into Perl's `Configure` file and figure out where the problem is. I tried to investigate that file but - yeah - things started getting messy. Nice to hear the reduce in size of `berkeley-db`.

**Comment by danielnachun** [comment]
@iMichka I see that the `perl` formula hasn't been fully ported to homebrew-core yet (it still references if `OS.mac?` and if `OS.linux?`).  Should I try to use this as an opportunity to fully backport it?  It looks fairly simple to me, but I don't know how that will behave when we try to merge back to linuxbrew-core.  

Otherwise I can revert the change you did in homebrew-core, and then I'd have to use a separate PR in linuxbrew-core to revert the changes that were unique to linuxbrew-core.

**Comment by danielnachun** [comment]
> Thanks, @danielnachun for digging into Perl's `Configure` file and figure out where the problem is. I tried to investigate that file but - yeah - things started getting messy and didn't have much time for further investigation. Nice to hear the reduce in size of `berkeley-db`.

This was definitely a case where we end up having to deal with design decisions made during an era where the primary concern was getting Perl to work on a bunch of different Unix platforms, not making it convenient to the install Perl in any location, possibly without administrative access, with fine-grained controls over dependencies.  I'm glad we've figured it out, and now that it's documented here hopefully it will save others the trouble of trying to figure it out in the future!

**Comment by iMichka** [comment]
> I see that the perl formula hasn't been fully ported to homebrew-core yet (it still references if OS.mac? and if OS.linux?). Should I try to use this as an opportunity to fully backport it? It looks fairly simple to me, but I don't know how that will behave when we try to merge back to linuxbrew-core.

Can you implement only the DB part? I am working on getting glibc shipped, which is needed by the perl postinstall step. Once this is done I'll do the remaining stuff.

**Comment by danielnachun** [comment]
Sounds good, I know what needs to be done for that.  I'll also make the PR to clean up BerkeleyDB so the bottle isn't so big.

**Comment by danielnachun** [comment]
The PRs to fix this now been merged and Perl builds correctly on my CentOS 7 system.  I'm going to close this as it should now be fixed, but feel free to follow up if something still isn't working.  @thiagogenez @ankitkk Thank you for your patience and assistance in figuring this out!

**Comment by tpmerrifield** [comment]
I'm getting this error when trying to install for the first time on RHEL7. I assumed it was related to the above issue when first encountered, so tried again when i saw this issue was closed. But still getting the same message about -ldb.
Using the install command listed on homepage:
```/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"```


```
==> Installing curl dependency: perl
==> ./Configure -des -Dprefix=~/linuxbrew/Cellar/perl/5.32.1 -Dprivlib=~/linuxbrew
Last 15 lines from ~/.cache/Homebrew/Logs/perl/01.Configure:
#include <stdio.h>
int main() { printf("Ok\n"); return(0); }

I used the command:

        cc -o try -O2 -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I~/linuxbrew/include -fstack-protector-strong -L/usr/local/lib try.c -lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
         ./try

and I got the following output:

~/linuxbrew/bin/ld: cannot find -ldb
collect2: error: ld returned 1 exit status
I can't compile the test program.
You have a BIG problem.  Shall I abort Configure [y]
Ok.  Stopping Configure.
```

I've tried both HEAD and master in the install command, likely doing something wrong, not sure what.
